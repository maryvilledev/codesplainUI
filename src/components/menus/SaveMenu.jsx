import React, { PropTypes } from 'react';
import {
  IconButton,
  IconMenu,
  MenuItem,
  Dialog,
  FlatButton,
  TextField,
} from 'material-ui';
import SaveIcon from 'material-ui/svg-icons/content/save';

import OrgSelector from '../OrgSelector';
import CustomPropTypes from '../../util/custom-prop-types';

const styles = {
  container: {
    height: '48px',
  },
};

const saveIconButton = disabled => (
  <IconButton
    disabled={disabled}
    title={disabled ? 'Login to Save' : 'Save Options'}
  >
    <SaveIcon />
  </IconButton>
);

class SaveMenu extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      saveAsName: '',
      showSaveAsDialog: false,
    };
    this.getMenu = this.getMenu.bind(this);
    this.getSaveAsDialog = this.getSaveAsDialog.bind(this);
  }

  getMenu() {
    const {
      canSave,
      enabled,
      onSaveClick,
    } = this.props;
    return (
      <IconMenu
        id="menu"
        iconButtonElement={saveIconButton(!enabled)}
      >
        <MenuItem
          onTouchTap={onSaveClick}
          disabled={!canSave}
          primaryText="Save"
        />
        <MenuItem
          onTouchTap={() => this.setState({ showSaveAsDialog: true })}
          primaryText="Save As"
        />
      </IconMenu>
    );
  }

  getSaveAsDialog() {
    const {
      onOrgChanged,
      orgs,
      selectedOrg,
    } = this.props;
    return (
      <Dialog
        actions={
          <span>
            <FlatButton
              label="Cancel"
              onTouchTap={() => this.setState({ showSaveAsDialog: false })}
            />
            <FlatButton
              primary
              label="Save"
              onTouchTap={() => {
                this.props.onSaveAsClick(this.state.saveAsName);
                this.setState({ showSaveAsDialog: false });
              }}
            />
          </span>
        }
        open={this.state.showSaveAsDialog}
        onRequestClose={() => this.setState({ showSaveAsDialog: false })}
      >
        <TextField
          floatingLabelText="Title"
          hintText="Enter new title"
          onChange={(ev, newVal) => this.setState({ saveAsName: newVal })}
        />
        <OrgSelector
          orgs={orgs}
          value={selectedOrg}
          onChange={onOrgChanged}
        />
      </Dialog>
    );
  }

  render() {
    return (
      <span style={styles.container}>
        {this.getMenu()}
        {this.getSaveAsDialog()}
      </span>
    );
  }
}

SaveMenu.propTypes = {
  canSave: PropTypes.bool.isRequired,
  enabled: PropTypes.bool.isRequired,
  onOrgChanged: PropTypes.func.isRequired,
  onSaveAsClick: PropTypes.func.isRequired,
  onSaveClick: PropTypes.func.isRequired,
  orgs: CustomPropTypes.orgs.isRequired,
  selectedOrg: PropTypes.string,
};

SaveMenu.defaultProps = {
  selectedOrg: '',
};

export default SaveMenu;
