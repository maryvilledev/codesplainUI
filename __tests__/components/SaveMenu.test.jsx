import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import getMuiTheme from 'material-ui/styles/getMuiTheme';

import SaveMenu from '../../src/components/menus/SaveMenu';

const defaultProps = {
  canSave: true,
  enabled: true,
  onSaveClick: jest.fn(),
  onSaveAsClick: jest.fn(),
};

describe('<SaveMenu />', () => {
  const muiTheme = getMuiTheme();
  const shallowWithContext = node => shallow(node, { context: { muiTheme } });
  describe('snapshot tests', () => {
    it('matches default snapshot', () => {
      const wrapper = shallowWithContext(
        <SaveMenu
          {...defaultProps}
        />,
      );
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
    it('matches snapshot when saving is not enabled', () => {
      const wrapper = shallowWithContext(
        <SaveMenu
          {...defaultProps}
          enabled={false}
        />,
      );
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
    it('matches snapshot when user cannot save', () => {
      const wrapper = shallowWithContext(
        <SaveMenu
          {...defaultProps}
          canSave={false}
        />,
      );
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
  });

  describe('prop: onSaveClick', () => {
    it('is is invoked when the "Save" MenuItem is clicked', () => {
      const onSaveClick = jest.fn();
      const wrapper = shallowWithContext(
        <SaveMenu
          {...defaultProps}
          onSaveClick={onSaveClick}
        />,
      );

      // Have to use 'touchTap' as opposed to 'click' here, see
      // "Common Gotchas" section of http://airbnb.io/enzyme/docs/api/ShallowWrapper/simulate.html
      wrapper.find('[primaryText="Save"]').simulate('touchTap');
      expect(onSaveClick.mock.calls.length).toBe(1);
    });
  });

  describe('the Save As dialog', () => {
    it('has a <TextField /> that updates the state when edited', () => {
      const wrapper = shallowWithContext(
        <SaveMenu {...defaultProps} />,
      );
      wrapper.find('TextField').simulate('change', {}, 'testing');
      expect(wrapper.instance().state.saveAsName).toBe('testing');
    });
    it('has a "Save" button that invokes the "onSaveAsClick" prop with the ' +
       'value of the "saveAsName" state field when clicked', () => {
      const onSaveAsClick = jest.fn();
      const wrapper = shallowWithContext(
        <SaveMenu
          {...defaultProps}
          onSaveAsClick={onSaveAsClick}
        />,
      );
      const title = 'test title';
      wrapper.instance().setState({ saveAsName: title });
      const dialogButtons = shallowWithContext(wrapper.find('Dialog').prop('actions'));
      dialogButtons.find('[label="Save"]').simulate('touchTap');
      expect(onSaveAsClick.mock.calls[0][0]).toBe(title);
    });
    it('has a "Cancel" button that does NOT invoke the "onSaveAsClick" prop ' +
       'when clicked', () => {
      const onSaveAsClick = jest.fn();
      const wrapper = shallowWithContext(
        <SaveMenu
          {...defaultProps}
          onSaveAsClick={onSaveAsClick}
        />,
      );
      const dialogButtons = shallowWithContext(wrapper.find('Dialog').prop('actions'));
      dialogButtons.find('[label="Cancel"]').simulate('touchTap');
      expect(onSaveAsClick.mock.calls.length).toBe(0);
    });
  });
});
