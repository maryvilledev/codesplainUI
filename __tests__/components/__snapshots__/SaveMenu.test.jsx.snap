exports[`<SaveMenu /> snapshot tests matches default snapshot 1`] = `
<span
  style={
    Object {
      "height": "48px",
    }
  }>
  <IconMenu
    anchorOrigin={
      Object {
        "horizontal": "left",
        "vertical": "top",
      }
    }
    animated={true}
    iconButtonElement={
      <IconButton
        disableTouchRipple={false}
        disabled={false}
        iconStyle={Object {}}
        title="Save Options"
        tooltipPosition="bottom-center"
        touch={false}>
        <ContentSave />
      </IconButton>
    }
    id="menu"
    multiple={false}
    onItemTouchTap={[Function]}
    onKeyboardFocus={[Function]}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onRequestChange={[Function]}
    onTouchTap={[Function]}
    open={null}
    targetOrigin={
      Object {
        "horizontal": "left",
        "vertical": "top",
      }
    }
    touchTapCloseDelay={200}
    useLayerForClickAway={false}>
    <MenuItem
      anchorOrigin={
        Object {
          "horizontal": "right",
          "vertical": "top",
        }
      }
      checked={false}
      desktop={false}
      disabled={false}
      focusState="none"
      insetChildren={false}
      onTouchTap={[Function]}
      primaryText="Save" />
    <MenuItem
      anchorOrigin={
        Object {
          "horizontal": "right",
          "vertical": "top",
        }
      }
      checked={false}
      desktop={false}
      disabled={false}
      focusState="none"
      insetChildren={false}
      onTouchTap={[Function]}
      primaryText="Save As" />
  </IconMenu>
  <Dialog
    actions={
      <span>
        <FlatButton
          disabled={false}
          fullWidth={false}
          label="Cancel"
          labelPosition="after"
          labelStyle={Object {}}
          onKeyboardFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onTouchStart={[Function]}
          onTouchTap={[Function]}
          primary={false}
          secondary={false} />
        <FlatButton
          disabled={false}
          fullWidth={false}
          label="Save"
          labelPosition="after"
          labelStyle={Object {}}
          onKeyboardFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onTouchStart={[Function]}
          onTouchTap={[Function]}
          primary={true}
          secondary={false} />
      </span>
    }
    autoDetectWindowHeight={true}
    autoScrollBodyContent={false}
    modal={false}
    onRequestClose={[Function]}
    open={false}
    repositionOnUpdate={true}>
    <TextField
      disabled={false}
      floatingLabelFixed={false}
      fullWidth={false}
      hintText="Enter new title"
      multiLine={false}
      onChange={[Function]}
      rows={1}
      type="text"
      underlineShow={true} />
  </Dialog>
</span>
`;

exports[`<SaveMenu /> snapshot tests matches snapshot when saving is not enabled 1`] = `
<span
  style={
    Object {
      "height": "48px",
    }
  }>
  <IconMenu
    anchorOrigin={
      Object {
        "horizontal": "left",
        "vertical": "top",
      }
    }
    animated={true}
    iconButtonElement={
      <IconButton
        disableTouchRipple={false}
        disabled={true}
        iconStyle={Object {}}
        title="Login to Save"
        tooltipPosition="bottom-center"
        touch={false}>
        <ContentSave />
      </IconButton>
    }
    id="menu"
    multiple={false}
    onItemTouchTap={[Function]}
    onKeyboardFocus={[Function]}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onRequestChange={[Function]}
    onTouchTap={[Function]}
    open={null}
    targetOrigin={
      Object {
        "horizontal": "left",
        "vertical": "top",
      }
    }
    touchTapCloseDelay={200}
    useLayerForClickAway={false}>
    <MenuItem
      anchorOrigin={
        Object {
          "horizontal": "right",
          "vertical": "top",
        }
      }
      checked={false}
      desktop={false}
      disabled={false}
      focusState="none"
      insetChildren={false}
      onTouchTap={[Function]}
      primaryText="Save" />
    <MenuItem
      anchorOrigin={
        Object {
          "horizontal": "right",
          "vertical": "top",
        }
      }
      checked={false}
      desktop={false}
      disabled={false}
      focusState="none"
      insetChildren={false}
      onTouchTap={[Function]}
      primaryText="Save As" />
  </IconMenu>
  <Dialog
    actions={
      <span>
        <FlatButton
          disabled={false}
          fullWidth={false}
          label="Cancel"
          labelPosition="after"
          labelStyle={Object {}}
          onKeyboardFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onTouchStart={[Function]}
          onTouchTap={[Function]}
          primary={false}
          secondary={false} />
        <FlatButton
          disabled={false}
          fullWidth={false}
          label="Save"
          labelPosition="after"
          labelStyle={Object {}}
          onKeyboardFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onTouchStart={[Function]}
          onTouchTap={[Function]}
          primary={true}
          secondary={false} />
      </span>
    }
    autoDetectWindowHeight={true}
    autoScrollBodyContent={false}
    modal={false}
    onRequestClose={[Function]}
    open={false}
    repositionOnUpdate={true}>
    <TextField
      disabled={false}
      floatingLabelFixed={false}
      fullWidth={false}
      hintText="Enter new title"
      multiLine={false}
      onChange={[Function]}
      rows={1}
      type="text"
      underlineShow={true} />
  </Dialog>
</span>
`;

exports[`<SaveMenu /> snapshot tests matches snapshot when user cannot save 1`] = `
<span
  style={
    Object {
      "height": "48px",
    }
  }>
  <IconMenu
    anchorOrigin={
      Object {
        "horizontal": "left",
        "vertical": "top",
      }
    }
    animated={true}
    iconButtonElement={
      <IconButton
        disableTouchRipple={false}
        disabled={false}
        iconStyle={Object {}}
        title="Save Options"
        tooltipPosition="bottom-center"
        touch={false}>
        <ContentSave />
      </IconButton>
    }
    id="menu"
    multiple={false}
    onItemTouchTap={[Function]}
    onKeyboardFocus={[Function]}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onRequestChange={[Function]}
    onTouchTap={[Function]}
    open={null}
    targetOrigin={
      Object {
        "horizontal": "left",
        "vertical": "top",
      }
    }
    touchTapCloseDelay={200}
    useLayerForClickAway={false}>
    <MenuItem
      anchorOrigin={
        Object {
          "horizontal": "right",
          "vertical": "top",
        }
      }
      checked={false}
      desktop={false}
      disabled={true}
      focusState="none"
      insetChildren={false}
      onTouchTap={[Function]}
      primaryText="Save" />
    <MenuItem
      anchorOrigin={
        Object {
          "horizontal": "right",
          "vertical": "top",
        }
      }
      checked={false}
      desktop={false}
      disabled={false}
      focusState="none"
      insetChildren={false}
      onTouchTap={[Function]}
      primaryText="Save As" />
  </IconMenu>
  <Dialog
    actions={
      <span>
        <FlatButton
          disabled={false}
          fullWidth={false}
          label="Cancel"
          labelPosition="after"
          labelStyle={Object {}}
          onKeyboardFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onTouchStart={[Function]}
          onTouchTap={[Function]}
          primary={false}
          secondary={false} />
        <FlatButton
          disabled={false}
          fullWidth={false}
          label="Save"
          labelPosition="after"
          labelStyle={Object {}}
          onKeyboardFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onTouchStart={[Function]}
          onTouchTap={[Function]}
          primary={true}
          secondary={false} />
      </span>
    }
    autoDetectWindowHeight={true}
    autoScrollBodyContent={false}
    modal={false}
    onRequestClose={[Function]}
    open={false}
    repositionOnUpdate={true}>
    <TextField
      disabled={false}
      floatingLabelFixed={false}
      fullWidth={false}
      hintText="Enter new title"
      multiLine={false}
      onChange={[Function]}
      rows={1}
      type="text"
      underlineShow={true} />
  </Dialog>
</span>
`;
